src/outdims.jl

```diff
@@ -0,0 +1,219 @@
+module NilNumber
+
+using LinearAlgebra
+using NNlib
+
+"""
+    Nil <: Number
+
+Nil is a singleton type with a single instance `nil`.
+Unlike `Nothing` and `Missing` it subtypes `Number`.
+"""
+struct Nil <: Number end
+
+const nil = Nil()
+
+Nil(::T) where T<:Number = nil
+(::Type{T})(::Nil) where T<:Number = nil
+
+Base.float(::Type{Nil}) = Nil
+Base.copy(::Nil) = nil
+Base.abs2(::Nil) = nil
+Base.sqrt(::Nil) = nil
+Base.zero(::Type{Nil}) = nil
+Base.one(::Type{Nil}) = nil
+
+Base.:+(::Nil) = nil
+Base.:-(::Nil) = nil
+
+Base.:+(::Nil, ::Nil) = nil
+Base.:+(::Nil, ::Number) = nil
+Base.:+(::Number, ::Nil) = nil
+
+Base.:-(::Nil, ::Nil) = nil
+Base.:-(::Nil, ::Number) = nil
+Base.:-(::Number, ::Nil) = nil
+
+Base.:*(::Nil, ::Nil) = nil
+Base.:*(::Nil, ::Number) = nil
+Base.:*(::Number, ::Nil) = nil
+
+Base.:/(::Nil, ::Nil) = nil
+Base.:/(::Nil, ::Number) = nil
+Base.:/(::Number, ::Nil) = nil
```

Comment:
With `Base.promote_rule(::Type{Nil}, ::Type{<:Number}) = Nil`, I think only the same-type version is needed:
```suggestion
for f in [:+, :-, :*, :/]
    @eval Base.$f(::Nil, ::Nil) = nil
end
```